activerecord/test/cases/connection_adapters/connection_handlers_multi_db_test.rb

```diff
@@ -120,6 +120,33 @@ def test_switching_connections_via_handler
           ENV["RAILS_ENV"] = previous_env
         end
 
+        def test_switching_connections_with_database_url
+          previous_env, ENV["RAILS_ENV"] = ENV["RAILS_ENV"], "default_env"
+          previous_url, ENV["DATABASE_URL"] = ENV["DATABASE_URL"], "postgres://localhost/foo"
+
+          ActiveRecord::Base.connected_to(database: "postgres://localhost/bar") do
+            handler = ActiveRecord::Base.connection_handler
+            assert_equal handler, ActiveRecord::Base.connection_handlers["postgres://localhost/bar"]
+          end
+        ensure
+          ActiveRecord::Base.establish_connection(:arunit)
+          ENV["RAILS_ENV"] = previous_env
+          ENV["DATABASE_URL"] = previous_url
+        end
+
+        def test_switching_connections_with_database_config
+          previous_env, ENV["RAILS_ENV"] = ENV["RAILS_ENV"], "default_env"
+          config = { "adapter" => "sqlite3", "database" => "db/readonly.sqlite3" }
+
+          ActiveRecord::Base.connected_to(database: config) do
+            handler = ActiveRecord::Base.connection_handler
+            assert_equal handler, ActiveRecord::Base.connection_handlers[config]
```

Comment:
I'm not sure how I feel about this. Should we try to lookup the config to store it as a hash key and fallback to this behaviour?

Reply:
Oh 💩 it just occurred to me this isn't going to work. It's fine if it's a symbol but what would the handler be if it was a url? We definitely don't want the handler to be `postgres://blah` so I think we need to change connects to to take a role and a database at the same time and only if it's a symbol use the database name as a handler. Thoughts?

Reply:
Yeah, I think if we always lookup handlers using symbols, that makes the most sense. Does it make sense to always use roles for handler lookup or do we need to still try using the database spec/env/config/url/??? for legacy support?

Reply:
> I've been thinking about this and I think we should either accept a symbol or a hash for the database option. If you need to pass a string then you need to also pass a role / name / handler with it.

I've refactored the method to behave this way for both URLs and config hashes. WDYT?